CONVENTIONS.md

```diff
@@ -0,0 +1,125 @@
+* Input arguments are always [`array_like`](https://numpy.org/doc/stable/glossary.html?highlight=array_like#term-array_like):
+    * "Any scalar or sequence that can be interpreted as an ndarray. In
+      addition to ndarrays and scalars this category includes lists
+      (possibly nested and with different element types) and tuples.
+      Any argument accepted by `np.array` is `array_like`."
+    * TBD: require a call to `np.asarray` or `np.asanyarray` on inputs?
+* All non-input arguments are keyword-only.
+* If `out` is `None` (the default), a new return array is created. Otherwise,
+  the result of the function is written to `out`.
+    * If there are multiple outputs, `out` is expected to be a tuple
+      with matching number of elements.
+* If `dtype` is `None` (the default), the dtype of the return array
+  will be determined automatically by numpy based on the operations
+  performed on the inputs. Otherwise, the dtype of the output array(s) is
+  overridden with the
+  provided dtype with a call to `ndarray.astype`. This should ensure
+  a matching precision of the calculation. The exact calculation
+  dtypes chosen may depend on the function and the inputs may be cast
+  to this dtype to perform the calculation.
+* There may be additional keyword-only arguments as required by
+  the specifics of individual functions.
+
+Since the conventions align with those of numpy, in some cases, it is
+possible to simply alias a numpy function to avoid all overhead and
+implement the conventions. Optionally `functools.partial`
+can be used to limit the available arguments, or a small wrapper function
+can be created.
+
+
+# Object oriented API conventions
+
+This API is for external use and for novice-users that want to
+perform linear algebra operations on conceptually familiar primitives.
+
+In this API each "thing" is represented as one object. 
+This API should make any linalg work much easier and safer, partly
+because semantics matters here: a point is not the same as a vector.
+
+These objects are array-like and iterable to make them easy to
+convert to native Python/Numpy objects. The objects support native python
+operators such as `__mul__` where applicable, and have methods
+specific to the type of object.
+
+Extensive input validation and ease of use is prioritized over performance.
+
+The source for this API resides in the `pylinalg.obj` subpackage and is organized
+by type.
+
+## Imports cycles
+
+Since the classes here will regularly need to instantiate other types,
+circular import dependencies exist. To work around this, only the `LinalgBase`
+type can be imported at module level, and other types will have to be imported
+at runtime in methods.
+
+## Function naming
+
+* Names should be concise and short.
+* For every method, there is an alternative in-place method, signified
+  with the prefix `i`. This is not the prettiest option, but it is
+  concise and short.
```

Comment:
This seems contradictory to the aim of "Extensive input validation and ease of use is prioritized over performance.". 

There is no upside to in-place computation other than performance and if you get it wrong you end up with really hard-to-debug problems which is not very beginner friendly. I would vote for excluding in-place computation and reserving it for the functional/advanced API (which can do it via `method(x1, out=x1)`).

Reply:
There is another upside; element-specific mutation detection can be implemented efficiently only with in-place mutations.

I'm inclined to agree with you but I'm having a hard time seeing how pygfx's `WorldObject` attributes `position` etc would work when they would need to be replaced all the time.

Reply:
Being able to assume that these objects are immutable also has favorable consequences (e.g. not having to worry about making a copy because the object might be modified without you knowing). But I agree that we need to see how this works out in PyGfx.

Reply:
Fair point. Could there be a middle ground by saying that objects are mutable in general, but that methods return copies? This would allow us to do in-place operations via the functional API (e.g. using `out=`) giving us access to performance while preserving the "validated and safe" aim of the object API.

I also like this because it makes it explicit where you could run into trouble: You are switching to the functional API here, so you need to pay extra attention to what you are doing.

Reply:
Illustrative example, I think:

```py
camera.position.z = 250
```

Would become:

```py
camera.position = camera.position + [0, 0, 250]
```

I don't think that's an improvement from the user perspective.